### PR TITLE
[Dygraph] Support sharding stage2/3+dp in GPT-3 model

### PR DESCRIPTION
**Pull request by haohongxiang**
### PR types

Bug fixes
### PR changes

Others
### Description

[Dygraph] Support sharding stage2/3+dp in GPT-3 model

一、GPT-3精度对比（单机8卡FP16训练1w后的loss）：
| strategy             | loss        |
| -------------------- | ----------- |
| dp8                  | 6.553567886 |
| sharding8(stage2)    | 6.553497314 |
| sharding8(stage3)    | 6.553690910 |
| dp2sharding4(stage2) | 6.553609848 |
| dp2sharding4(stage3) | 6.553627491 |

二、适配Demo
```python
import paddle
from paddle.fluid.dygraph.nn import Linear
from paddle.distributed import fleet
from paddle.vision.models import ResNet
from paddle.vision.models.resnet import BasicBlock

###### 1.引入依赖包
from paddle.distributed.sharding import group_sharded_parallel
from paddle.fluid.dygraph.parallel import sync_params_buffers
from paddle.distributed.fleet.utils.hybrid_parallel_util import fused_allreduce_gradients
######

sharding_stage = 3 # or 2
dp_degree = 2
mp_degree = 1
pp_degree = 1
sharding_degree = 4
use_pure_fp16 = True
level = "p_g_os" if sharding_stage == 3 else "os_g"

###### 2.初始化分布式环境
strategy = fleet.DistributedStrategy()
strategy.hybrid_configs = {
    "dp_degree": dp_degree,
    "mp_degree": mp_degree,
    "pp_degree": pp_degree,
    "sharding_degree": sharding_degree
}
fleet.init(is_collective=True, strategy=strategy)
hcg = fleet.get_hybrid_communicate_group()
dp_group = hcg.get_data_parallel_group()
sharding_group = hcg.get_sharding_parallel_group()
######

if sharding_stage in [2, 3]:
    assert mp_degree == pp_degree == 1, "sharding stage2/3 will support tensor/pipeline parallel later"

model = ResNet(BasicBlock, 18)
clip = paddle.nn.ClipGradByGlobalNorm(clip_norm=1.0)
optimizer = paddle.optimizer.AdamW(learning_rate=0.001, parameters=model.parameters(), weight_decay=0.00001, grad_clip=clip)
scaler = None

if use_pure_fp16:
    scaler = paddle.amp.GradScaler(init_loss_scaling=32768)
    model = paddle.amp.decorate(
        models=model,
        level='O2',
        save_dtype='float32')

###### 3.dp通信组内广播参数，然后用sharding api封装
if dp_degree > 1:
    sync_params_buffers(model, comm_group=dp_group, src_rank=dp_group.ranks[0])

model, optimizer, scaler = group_sharded_parallel(model, optimizer, level=level, scaler=scaler, group=sharding_group)
######
# 如何继续训练？

for step_id in range(1, 100):
    x = paddle.rand([1, 3, 224, 224])
    with paddle.amp.auto_cast(
        use_pure_fp16, custom_black_list=[
            "reduce_sum",
            "c_softmax_with_cross_entropy",
            "elementwise_div"], level='O2'):
        out = model(x)
    loss = out.mean()
    if use_pure_fp16:
        scaler.scale(loss).backward()
    else:
        loss.backward()
    
    print("=== step_id : {}    loss : {}".format(step_id, loss.numpy()))

    ###### 4.优化器计算前在dp通信组内进行梯度同步
    if sharding_stage in [2, 3] and dp_degree > 1:
        fused_allreduce_gradients(model.parameters(), hcg)
        if sharding_stage == 3:
            for p in model.parameters():
                if hasattr(p, "bw_storage"):
                    assert p.grad is None, "This case shouldn't happen."
                    p.bw_storage.scale_(1.0 / dp_group.nranks)
                    paddle.distributed.all_reduce(p.bw_storage, group=dp_group)
    ######

    if use_pure_fp16:
        scaler.step(optimizer)
        scaler.update()
    else:
        optimizer.step()
    optimizer.clear_grad()

# stage3: param save
# stageX: optimizer. save
# Load。 group_shard 之前load.

```

### CODE REVIEW

**Review by ForFishes — DISMISSED** [review]
LGTM

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 1]
```diff
@@ -353,6 +360,16 @@ def do_train(args):
                             loss_mbs.backward()
                         loss = loss + loss_mbs
 
+                    if args.sharding_stage in [2, 3] and args.dp_degree > 1:
+                        fused_allreduce_gradients(model.parameters(), hcg)
+                        if args.sharding_stage == 3:
+                            for p in model.parameters():
+                                if hasattr(p, "bw_storage"):
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py:374]: 这个 `bw_storage` 是？

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 2]
```diff
@@ -353,6 +360,16 @@ def do_train(args):
                             loss_mbs.backward()
                         loss = loss + loss_mbs
 
+                    if args.sharding_stage in [2, 3] and args.dp_degree > 1:
+                        fused_allreduce_gradients(model.parameters(), hcg)
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py:371]: 这里`fused_allreduce_gradient`和下面都加一下注释吧。不清楚是做了啥通信

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 3]
```diff
@@ -272,6 +275,10 @@ def do_train(args):
     # wrap sharding stage2/3 and add collective group
     # TODO(Baibaifan): combine ShardingStage1/2/3 and fleet.distributed_model in feature
     if args.sharding_stage in [2, 3]:
+        if args.dp_degree > 1:
+            sync_params_buffers(
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py]: 加注释

**Comment by ZHUI** [comment]
cc: @haohongxiang 这个pr还有效吗，有效的话我合入了

**Comment by haohongxiang** [comment]
> cc: @haohongxiang 这个pr还有效吗，有效的话我合入了

@ZHUI 嗯嗯，还有效，可以合入

**Comment by ZHUI** [comment]
> > cc: @haohongxiang 这个pr还有效吗，有效的话我合入了
> 
> @ZHUI 嗯嗯，还有效，可以合入

辛苦重新commit 修改一下 CodeStyle，我来合入

**Comment by haohongxiang** [comment]
> > > cc: @haohongxiang 这个pr还有效吗，有效的话我合入了
> > 
> > 
> > @ZHUI 嗯嗯，还有效，可以合入
> 
> 辛苦重新commit 修改一下 CodeStyle，我来合入

@ZHUI 可以合入了

**Review by gongel — APPROVED** [review]
